plasmapy/utils/code_repr.py

```diff
@@ -0,0 +1,257 @@
+            elif power == 1:
+                formatted += f"*u.{base}"
+            else:
+                formatted += f"*u.{base}**{power}"
+
+    return formatted
+
+
+def _format_arg(arg) -> str:
+    """Transform an argument into a format as would appear in a function call."""
+    if hasattr(arg, "__name__"):
+        return arg.__name__
+    elif isinstance(arg, u.Quantity):
+        return _format_quantity(arg)
+    else:
+        return repr(arg)
+
+
+def _format_kw(keyword) -> str:
+    """Transform a keyword into a format as would appear in a function call."""
+    if isinstance(keyword, str):
+        return str(keyword)
+    elif hasattr(keyword, "__name__"):
+        return keyword.__name__
+    else:
+        return repr(keyword)
+
+
+def _format_args_and_kwargs(args: Any = tuple(), kwargs: Dict = {}) -> str:
+    """
+    Take positional and keyword arguments, and format them into a
+    string as they would appear in a function call.
+    """
+    args = args if isinstance(args, tuple) else (args,)
+    args_and_kwargs = ""
+
+    for arg in args:
+        args_and_kwargs += f"{_format_arg(arg)}, "
+
+    for kwarg in kwargs:
+        args_and_kwargs += f"{_format_kw(kwarg)}={_format_arg(kwargs[kwarg])}, "
+
+    if args_and_kwargs[-2:] == ", ":
+        args_and_kwargs = args_and_kwargs[:-2]
+
+    return args_and_kwargs
+
+
+def _name_with_article(ex: Exception) -> str:
+    """
+    Return a string with an indefinite article and the name of
+    exception ``ex``.
+    """
+    exception_name = ex.__name__
+    use_an = exception_name[0] in "aeiouAEIOU" and exception_name[0:4] != "User"
+    indefinite_article = "an" if use_an else "a"
+    return f"{indefinite_article} {exception_name}"
+
+
+def _object_name(obj: Any, showmodule=False) -> str:
+    """
+    Return the name of an `object`.  If the `object` has a `__name__`
+    attribute and ``showmodule`` is `True`, then prepend the module
+    name if not in `builtins`.
+    """
+    obj_name = obj.__name__ if hasattr(obj, "__name__") else repr(obj)
+
+    if hasattr(obj, "__name__") and showmodule is True:
+        module_name = inspect.getmodule(obj).__name__  # ADD TEST
+        if module_name != "builtins":
+            obj_name = f"{module_name}.{obj_name}"  # ADD TEST
+
+    return obj_name
+
+
+def _string_together_warnings_for_printing(
+    warning_types: List[Warning], warning_messages: List[str]
+):
+    """
+    Take a list of warning types with a list of corresponding warning
+    messages, and create a string that prints out each warning type
+    followed by the corresponding message, separated by a full line.
+    """
+    warnings_with_messages = [
+        _object_name(warning, showmodule=False) + ": " + message
+        for warning, message in zip(warning_types, warning_messages)
+    ]
+    return "\n\n".join(warnings_with_messages)
+
+
+def call_string(f: Callable, args: Any = tuple(), kwargs: Dict = {}) -> str:
```

Comment:
I'm trying to understand the purpose of `call_string`.  Is it used for actual `assert` tests or is used only for reporting info associated with test failures?  For example,

```python
pytest.fail(f"Call {cal_string(...)} was suppose to do blah, but did doogah.")
```

I don't know, this seems like a lot of complexity and we're traveling rather far from [KISS](https://en.wikipedia.org/wiki/KISS_principle).  This functionality works beautifully for simple function calls, but would get out of hand if you were trying to use it on a function with a signature like [`scipy.signal.find_peaks`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html) (random example that would plausible in a `plasmapy.analysis` function).

Reply:
The main purpose of `call_string`, etc. is for reporting info associated with test failures.  Reading the actual function call that led to a test failure can be extremely helpful for figuring out what the problem is, and it can help speed up debugging too, especially for newcomers to the project.  These functions could also be used in `__repr__`, and I think there would likely be times when these could be used directly in tests too.  

The point you made about managing complexity is important.  I'll need to think about that more...

Reply:
I'm curious if you could sidestep the complexity if you incorporate some functionality form `inspect` to do the heavy lifting, like [`getcallargs`](https://docs.python.org/3/library/inspect.html#inspect.getcallargs) in combination with [`formatargspec`](https://docs.python.org/3/library/inspect.html#inspect.formatargspec).

Reply:
That looks promising!  I'll look into it!

Reply:
I'm not sure if what's in `inspect` will be able to take care of all the use cases that are needed.  Since `call_string` is being used to report testing information, it needs to be able to handle situations where the call is invalid, since invalid calls will sometimes need to be reconstructed and reported back to the user.  That is apparently out of scope for `inspect`.  Hm...